beancount_import/source/generic_importer_source.py

```diff
@@ -0,0 +1,148 @@
+"""This module implements a Source Subclass for wrapping
+`beancount.ingest.importer.ImporterProtocol` subclasses importers.
+The importers are considered athoritative of the account they represent.
+
+The Transaction.narration set by each importer is copied to Posting.meta[source_desc]
+This helps in predicting postings for similar transaction while allowing the
+user to change the Transaction description and payee from UI (see readme.md for more on source_desc)
+
+Author: Sufiyan Adhikari(github.com/dumbPy)
+"""
+
+import os
+import hashlib
+from glob import glob
+from typing import List, Tuple
+
+from beancount.core.data import Transaction, Posting,  Directive
+from beancount.core.amount import Amount
+from beancount.ingest.importer import ImporterProtocol
+from beancount.core.compare import hash_entry
+from beancount.ingest.cache import get_file
+from beancount.ingest.similar import find_similar_entries, SimilarityComparator
+
+from ..matching import FIXME_ACCOUNT, SimpleInventory
+from . import ImportResult, Source, SourceResults, InvalidSourceReference, AssociatedData
+from ..journal_editor import JournalEditor
+
+
+class ImporterSource(Source):
+    def __init__(self,
+                 directory: str,
+                 account: str ,
+                 importer: ImporterProtocol,
+                 account_name:str = None,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.directory = os.path.expanduser(directory)
+        self.importer = importer
+        self.account = account
+        self.account_name = account_name if account_name else self.name
+
+        self.comparator = SimilarityComparator()
+
+        # get _FileMemo object for each file
+        files = [get_file(f) for f in
+                 glob(os.path.join(directory, '**', '*'), recursive=True)
+        ]
+        # filter the valid files for this importer
+        self.files = [f for f in files if self.importer.identify(f)]
+
+    @property
+    def name(self):
+        return self.importer.name()
+
+    def prepare(self, journal: 'JournalEditor', results: SourceResults) -> None:
+        results.add_account(self.account)
+        entries = {}
+        for f in self.files:
+            f_entries = self.importer.extract(f)
+            # deduplicate across statements
+            hashed_entries = {}
+            for entry in f_entries:
+                hash_ = self._hash_entry(entry, frozenset(['filename','lineno']))
```

Comment:
Sometimes you can end up with two completely identical transactions, that correspond to two different real transactions, if the datasource doesn't provide a unique transaction id, just a description (e.g. two transactions in the same amount on the same day with the same merchant).  With the type of deduplication you have here, all but one of those transactions will be silently discarded.

The way I've dealt with this in the other data sources is to count for a given source transaction the number of copies in the source data, and expect the same number of copies in the journal.

Reply:
I actually have a this in my bank statements. Will add hash->count mapping  and use that. Thank you!

Reply:
Fixed. Now I am maintaining hash->List[Transaction] mapping to cater to this scenario.